Define gem-changelog-update-check action

MatthewPickinYork opened the pull request:
# Context
See FD-1417; basically, it's an action for checking if a) a `CHANGELOG.md` file exists, and if it does, then b) checking if it's been changed if anything in `lib/` has been changed.

# Implementation
Pretty straight-forward, just a bit of bash to make the check; it does first need to establish what files have changed, and then just grep for the relevant files.

MatthewPickinYork committed:
Define gem-changelog-update-check action

MatthewPickinYork committed:
Allow for the changelog file path as an input

MatthewPickinYork committed:
Update README with gem-changelog-update-check action info

MatthewPickinYork committed:
Add debug statements to action

MatthewPickinYork committed:
Fix debug statement

njm506 commented on gem-changelog-update-check/action.yml:
```diff
@@ -0,0 +1,25 @@
+        	echo $changes | grep -q -E 'lib/.+' && \
+        	echo $changes | grep -q '${{ inputs.changelog-file-path }}'
+        else
+        	echo "No CHANGELOG.md to update"
```
Use `inputs.changelog-file-path` here.  Maybe set it as a shell variable within the `run` block.

njm506 commented on gem-changelog-update-check/action.yml:
```diff
@@ -0,0 +1,25 @@
+    - uses: actions/checkout@v4
+    - name: Check if CHANGELOG needs updating
+      run: |
+        changes=$(git diff --name-only --diff-filter=ACMRT ${{ github.event.pull_request.base.sha }} ${{ github.sha }})
```
Always surround `$()` with `""` (in assignment it may not matter but word splitting is important.

njm506 commented on gem-changelog-update-check/action.yml:
```diff
@@ -0,0 +1,25 @@
+    - name: Check if CHANGELOG needs updating
+      run: |
+        changes=$(git diff --name-only --diff-filter=ACMRT ${{ github.event.pull_request.base.sha }} ${{ github.sha }})
+        echo $changes
```
`""` around `$changes` to prevent word splitting

njm506 commented on gem-changelog-update-check/action.yml:
```diff
@@ -0,0 +1,25 @@
+        if [ -f "${{ inputs.changelog-file-path }}" ]; then
+            echo $changes &&
+        	echo $changes | grep -q -E 'lib/.+' && \
+        	echo $changes | grep -q '${{ inputs.changelog-file-path }}'
```
You're echoing changes here while it's already done above.  Again, quoting.

Use an anchor with the `lib/` pattern so as not to match e.g. `public/lib/` or `flib/` - probably just `^lib/`.  You probably don't need `.+` because you don't care about anything after `lib/`, just that things under there changed.

Use `fgrep -x` (`grep -F -x`) to exactly match the entire line for the changelog file path.

MatthewPickinYork replied:
Ah, so this part is what I'm working on. I originally did use `^lib/.*+$` as my regex, but I was finding the output of the `git diff` command to spit out the changes to a space-separated list rather than newline, which threw things off.

Regardless, the `git diff` command I'm using _isn't actually working_ (it's not returning anything), so that's what I'm currently looking into.

njm506 replied:
By default, git diff tells you changes to the working tree, you might want to be diffing against the main branch (maybe there's an input for PR merge target)?

You can also `git diff` particular subtrees, e.g. `git diff main -- /lib` will only diff the `lib` directory against main, and you can do the same with `CHANGELOG.md` I think, so maybe that is easier than filtering its output.  I suspect the reason you're seeing spaces is that $changes isn't quoted, so the command splitting is removing all whitespace and turning it into spaces.

njm506 replied:
It may also need to inspect the .gemspec file, since that is used to build the output (is there something in the rakefile that figures out what to include, or is it just "everything")?

MatthewPickinYork replied:
I think the rakefile leverages `bundler/gem_tasks` to handle a lot of the gem-related tasks, and so it doesn't really surface much in terms of e.g. tasks to show what stuff is changed.

How about `'^lib/.+$||^.*\.gemspec$'`?

njm506 replied:
I mean it's better than what it is now.

We probably should include spec changes too...

So `'^lib/|^spec/|\.gemspec$'` is probably OK (you don't need to have the whole line match the regex) - probably put it in a shell variable near the top of the script so it's obvious though.

MatthewPickinYork replied:
I think I'll throw `^bin/` in there too

njm506 reviewed (changes requested):
This is good but there are a couple of things in the regexes that could be improved.

Also, your code would be best run through shellcheck.  The easiest way to achieve that is to put it in a separate `.sh` file (where tools can be run against it) and then invoke it from the `run`, passing inputs explicitly.

MatthewPickinYork committed:
Add further debug statements

MatthewPickinYork committed:
Add git status command to check status of checked-out repo

MatthewPickinYork committed:
Hard-code main branch to git diff against

MatthewPickinYork committed:
Remove debug statements, shellcheck recommended changes

- Remove unneeded debug echo commands
- Enclose variables (assignment and calls) in quotes

MatthewPickinYork committed:
Interpolate name of changelog instead of hard-coding it

MatthewPickinYork committed:
Move changelog path to bash variable

- To simplify run script, assign changelog path to
  variable
- Use double quotes where string interpolation is
  required

MatthewPickinYork committed:
Add fetch-depth:0 parameter to checkout action

- Explicitly fetch entire git history via fetch-depth:0
  parameter
- Check changes relative to main branch

MatthewPickinYork committed:
Add print statement to show changed files via git diff

MatthewPickinYork committed:
Further specify branches to compare via git diff

Ensure current HEAD is compared to upstream origin
main branch

MatthewPickinYork committed:
Refine regex for identifying (changed) lib/ files

MatthewPickinYork committed:
Add more patterns to file regex, move to shell variable

njm506 commented on gem-changelog-update-check/action.yml:
```diff
@@ -0,0 +1,29 @@
+        echo "Changes relative to main branch:"
+        echo "$changes"
+        if [ -f "$changelog_file" ]; then
+        	echo "$changes" | grep -q -E "$file_regex" && \
+        	echo "$changes" | grep -q "$changelog_file"
```
Use `fgrep -qx` or `grep -Fqx` to search for the exact line "CHANGELOG.md" (or whatever was specified).

njm506 replied:
Use two-space soft tabs.

njm506 commented on gem-changelog-update-check/action.yml:
```diff
@@ -0,0 +1,29 @@
+        changelog_file="${{ inputs.changelog-file-path }}"
+        echo "Changes relative to main branch:"
+        echo "$changes"
+        if [ -f "$changelog_file" ]; then
```
In shell scripts, I'd avoid using `; then`; put it on a separate line (same with `do` in a `while` loop)
```
if [ ... ]
then
```

njm506 commented on readme.md:
````diff
@@ -126,6 +126,28 @@ jobs:
+```yaml
+jobs:
+  gem-changelog-update-check:
+    name: Gemfury Deployment
````
Name is wrong here :)

MatthewPickinYork replied:
Ooops :/

njm506 reviewed (commented):
Broadly fine but a couple of things to address.

MatthewPickinYork committed:
Correct README using the wrong name parameter

MatthewPickinYork committed:
Refine grep for searching for changelog file

- Use fgrep functionality when searching for
  changelog file in list of file changes
- Style fixes (two-space soft tabs, if statement
  sans 'then' keyword

MatthewPickinYork committed:
Add missing then keyword to bash if statement

njm506 commented on gem-changelog-update-check/action.yml:
```diff
@@ -0,0 +1,30 @@
+    - name: Check if CHANGELOG needs updating
+      run: |
+        file_regex='^lib/|^bin/|^spec/|\.gemspec$'
+        changes="$(git diff --name-only --diff-filter=ACMRT origin/main HEAD)"
```
`--diff-filter` should probably include **D**eleted too; if you delete something from `lib` there should be an explanation in the changelog, even if it's a refactor.

Furthermore, do we need `--diff-filter` at all?  If anything changes in those files/dirs, CHANGELOG should be updated, no?

MatthewPickinYork replied:
I've gone with removing `--diff-filter` completely - I agree, we want as broad search for changes as possible

MatthewPickinYork committed:
Remove diff-filter flag and parameters

As we're looking for any changes to the listed files, remove
the filter to broaden the search space

njm506 commented on gem-changelog-update-check/action.yml:
```diff
@@ -0,0 +1,30 @@
+        fetch-depth: 0
+    - name: Check if CHANGELOG needs updating
+      run: |
+        file_regex='^lib/|^bin/|^spec/|\.gemspec$'
```
Suggest structuring the regex:
```
file_regex='^(bin|lib|spec)/|\.gemspec$'
```

You could always bracket gemspec too if you want but we only look for one file extension at the moment.

njm506 commented on gem-changelog-update-check/action.yml:
```diff
@@ -0,0 +1,30 @@
+        if [ -f "$changelog_file" ]
+        then
+          echo "$changes" | grep -q -E "$file_regex" && \
+          echo "$changes" | grep -Fqx "$changelog_file"
```
Combine your flags in a consistent manner; if you're going to use `-Fqx` style on one line, use `-Eq` on the other.

njm506 commented on gem-changelog-update-check/action.yml:
```diff
@@ -0,0 +1,30 @@
+    - uses: actions/checkout@v4
+      with:
+        fetch-depth: 0
+    - name: Check if CHANGELOG needs updating
```
Check that changelog was updated if necessary

njm506 commented on readme.md:
```diff
@@ -126,6 +126,28 @@ jobs:
+## gem-changelog-update-check
+
+Used for pull requests to check if the CHANGELOG.md (or specified file) has been updated if files in the `lib/`
+directory have been updated. The test will return a fail if changes are required but haven't been made.
```
...if the changelog... (don't need to name it with caveat, it's described in the inputs section)
...if the gemspec, or files in lib/ bin/ or spec/, have been updated.

njm506 commented on gem-changelog-update-check/action.yml:
Am I correct in thinking that exit status indicates whether the check passes or fails?

If that is the case, I would recommend adding `exit 0` and `exit 1` to the shell script since the exit status isn't immediately obvious because of how the logic is structured.

I'd suggest changing the "no changelog file exists" test to a guard, i.e. message, `exit 0` early and get out of the conditional.  That makes the fail-to-ok behaviour explicit.

The remaining part of the script would need to check that the changelog file has been modified (causes the check to pass), or that no relevant files were modified.  You can always use `elif` to do each check in sequence.

MatthewPickinYork replied:
I've given this a go; it's bumped up the number of lines, but I'm hoping it's more clear in what it's doing.

njm506 reviewed (commented):
I'm broadly happy with this but I think a slightly different structure in the script might be more easily understood by others.

MatthewPickinYork committed:
Structure regex term for improved readability

MatthewPickinYork committed:
Ensure command flag combos are consistent across the script

MatthewPickinYork committed:
Update job step name for improved readability

MatthewPickinYork committed:
Update README with a more accurate description of the new action

MatthewPickinYork committed:
Refactor script for improved readability

Use explicit exit commands and guard statements to
improve readability.

njm506 commented on gem-changelog-update-check/action.yml:
```diff
@@ -0,0 +1,40 @@
+            echo "$changelog_file has not been updated" >&2
+            exit 1
+          fi
+        fi
```
I'd reorder this and use `elif` so that you avoid nested branching paths (pseudocode):
```
if(no_changelog) => ok (no changelog to update)
elif(changelog_updated) => ok (changelog was updated; doesn't matter whether anything changed)
elif(relevant_changes) => fail (files changed but changelog wasn't updated)
else => ok (no relevant files changed)
```

Arguably if you use a structure like this, you could put the exit 0 at the end (for completeness), and only have the exit 1 within the logic.

Some blank lines to pad things out might make things clearer (although I'm not sure if the inline YAML permits that...)

MatthewPickinYork replied:
For some reason I completely discounted the fact that having just the changelog change would be fine - it does make good sense that one should be able to do this!

I've left in the explicit `exit 0` statements within the branches, as I'm not sure how to otherwise structure things. Also, I think it's more clear this way.

MatthewPickinYork committed:
Restructure branching logic

njm506 commented on gem-changelog-update-check/action.yml:
```diff
@@ -0,0 +1,36 @@
+          exit 0
+        elif echo "$changes" | grep -Fqx "$changelog_file"
+        then
+          exit 0
```
I think it's still useful to log here with an `echo "$changelog_file is up to date"`

MatthewPickinYork committed:
Add printout to stdout when changes to changlog detected
